IO/BeagleBone/CANBusBBB.cs

```diff
@@ -3,56 +3,160 @@
+		}
+
+		/// <summary> Converts a pin number to the corresponding CAN bus ID. 255 if invalid. </summary>
+		static internal byte PinToCANBus(BBBPin Pin)
+		{
+			switch (Pin)
+			{
+				case BBBPin.P9_19:
+				case BBBPin.P9_20: return 0;
+
+				case BBBPin.P9_24:
+				case BBBPin.P9_26: return 1;
+			}
+			return 255;
+		}
+	}
+
+	public class CANBusBBB : ICANBus
+	{
+		[StructLayout(LayoutKind.Sequential)]
+		private struct SockAddrCan
+		{
+			[MarshalAs(UnmanagedType.U2)]
+			public ushort CanFamily;
+
+			[MarshalAs(UnmanagedType.I4)]
+			public int CanIfIndex;
+
+			[MarshalAs(UnmanagedType.U4)]
+			public uint RxId;
+
+			[MarshalAs(UnmanagedType.U4)]
+			public uint TxId;
+		}
+
+		[StructLayout(LayoutKind.Explicit)]
+		private unsafe struct IfRequest
+		{
+			[FieldOffset(0), MarshalAs(UnmanagedType.ByValArray, SizeConst = 16)]
+			public fixed byte Name[16];
+
+			[FieldOffset(16), MarshalAs(UnmanagedType.I4)]
+			public int IfIndex;
+
+			[FieldOffset(16), MarshalAs(UnmanagedType.ByValArray, SizeConst = 24)]
+			public fixed byte IfrU[24];
+		}
+
+		[StructLayout(LayoutKind.Explicit)]
+		private unsafe struct CANFrame
+		{
+			[FieldOffset(0), MarshalAs(UnmanagedType.U4)]
+			public uint CANID;
+
+			[FieldOffset(4), MarshalAs(UnmanagedType.U1)]
+			public byte DataLength;
+
+			[FieldOffset(5), MarshalAs(UnmanagedType.U1)]
+			public byte Padding;
+
+			[FieldOffset(6), MarshalAs(UnmanagedType.U1)]
+			public byte Reserved0;
+
+			[FieldOffset(7), MarshalAs(UnmanagedType.U1)]
+			public byte Reserved1;
+
+			//The original source forced this to 8-byte alignment with __attribute__((aligned(8)))
+			//If something breaks, maybe fix it?
+			[FieldOffset(8), MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
+			public fixed byte Data[8];
+		}
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int socket(int Namespace, int Style, int Protocol);
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int ioctl(int FileDescriptor, uint Request, ref IfRequest Req);
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int bind(int FileDescriptor, ref SockAddrCan Addr, int AddrLen);
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int read(int FileDescriptor, ref CANFrame Frame, int Size);
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int write(int FileDescriptor, ref CANFrame Frame, int Size);
+
+		[DllImport("libc", SetLastError = true)]
+		private static extern int close(int FileDescriptor);
+
+		private const int AF_CAN = 29;
+		private const int PF_CAN = AF_CAN;
+		private const int SOCK_RAW = 3;
+		private const int CAN_RAW = 1;
+		private const uint SIOCGIFINDEX = 0x8933;
+		private int Socket;
+
+		internal CANBusBBB(string CanName)
+		{
+			this.Socket = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+			if (this.Socket < 0) { throw new Exception("Error while opening socket. Error code: " + Marshal.GetLastWin32Error()); }
+
+			unsafe
```

Comment:
Can we limit the scope of this `unsafe` block? Right now it's wrapping most of the method. If only a couple of lines actually need it, it'd probably be best to reduce the size of the block as much as possible.

Reply:
`unsafe` block cannot be shrinked because `IfRequest` is `unsafe`, and is used on the first and second-to-last line of the block.